style_guide.md

```diff
@@ -0,0 +1,130 @@
+
+## Common Jsonnet
+
+### Lib folders
+
+- Rules and placeholders are located in `lib` folder in corresponding schemas along with `blocks` folder.
+
+### Rules and placeholders
+
+- These are separated into: `common_rules`, `placeholders` and `rules`. These files should end with `.libsonnet` suffix.
+
+### Top level and extended
+
+- `common_rules` and `placeholders` are located in top-level "lib" folder.
+- `rules` files for each schema are located in schema's sub-folder and extend some of the top-level `common_rules`.
+
+## Global variables
+
+### Top level arguments
+
+- For each schema created it is possible to pass variables, e.g.: `region_code` and `census_month_year_date`, with different outputs being generated depending on their value.
+- These are defined in eq-questionnaire-runner, are known at questionnaire launch and won't change during the Census.
+
+### External variables
+
+- External variables are being used. Example of that is `census_month_year_date`.
+
+### Region Code
+
+- Depending on `region_code` variable's value, different structure of sections is generated (e.g. certain questions omitted).
+
+## Blocks
+
+### Structure
+
+#### Imports
+
+- They should be located at the top of the file.
+- Line length limit doesn't apply.
+
+#### Variables
+
+- Variables should be defined according to general principles and styling rules.
+
+#### Methods
+
+- We should define a method each time we’re changing a property, e.g. in individual→blocks→identity-and-health→carer.jsonnet:
+    - question should be called with question(isProxy=true) or question(isProxy=false)
+    - In the question method, when we need to resolve title, we call a new questionTitle method, passing isProxy
+    - In the questionTitle method we would return either of the two strings, based on the isProxy variable
+    - We would then do a similar thing for guidance with a questionGuidance method
+- Methods can be invoked at top-level or declared inline, depending on the intended use:
+```
+    // TOP-LEVEL
+    
+    local proxyTitle = {
+      text: 'Does <em>{person_name}</em> mainly work in the UK?',
+      placeholders: [placeholders.personName,],
+    };,
+    
+    // INLINE
+    
+    contents: [
+      {
+        description: {
+          text: regionDescriptionProxy,
+          placeholders: [placeholders.personName],
+      },
+    },
+  ```
+
+### Block json
+
+- Block json can be a function if an argument is being passed.
+
+### Variants
+
+- Where variants exist, a question method should be used and passed the variables that control the variant, e.g. `isProxy`.
+```
+    {
+      type: 'Question',
+      id: 'date-of-birth',
+      question_variants: [
+        {
+          question: question(nonProxyTitle),
+          when: [rules.isNotProxy],
+        },
+        {
+          question: question(proxyTitle),
+          when: [rules.isProxy],
+        },
+      ],
+    }
+```
```

Comment:
I think we should move away from passing the title etc to the methods as it's led to a proliferation of things being passed when more than one thing varies. We should pass the boolean that the variant is based on e.g. here it would be `isProxy=true` or `isProxy=false`.

Reply:
Changed it to `isProxy`, thanks!